src/rrd/mod.rs

```diff
@@ -0,0 +1,493 @@
+    window: i64,
+    retention: i64,
+}
+
+pub struct RRDSlotImpl {
+    connection: sqlx::pool::PoolConnection<Sqlite>,
+    key: i64,
+}
+
+#[async_trait]
+impl Slot for RRDSlotImpl {
+    async fn counter(&mut self, key: &str, value: f64) -> Result<()> {
+        let last = RRDSlotImpl::get_last(self, key).await?;
+        RRDSlotImpl::set_last(self, key, &value).await?;
+        if last.is_none() {
+            return Ok(());
+        }
+        let last = last.unwrap();
+        let diff = if value >= last {
+            value - last
+        } else {
+            // this is either an overflow
+            // or counter has been reset (node was restarted hence)
+            // metrics are counting from 0 again.
+            // hence it's safer to assume diff is just the value
+            // reported
+            value
+        };
+        sqlx::query("REPLACE INTO usage (timestamp, metric, value) VALUES (?, ?, ?);")
+            .bind(self.key)
+            .bind(key)
+            .bind(diff)
+            .execute(&mut self.connection)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn key(&mut self) -> Result<i64> {
+        Ok(self.key)
+    }
+}
+
+#[async_trait]
+impl RRD<RRDSlotImpl> for RRDImpl {
+    async fn last(&mut self, metric: &str) -> Result<Option<f64>> {
+        let mut slot = RRDImpl::slot(self).await?;
+        Ok(slot.get_last(metric).await?)
+    }
+
+    async fn counters(&mut self, since: std::time::SystemTime) -> Result<Vec<(String, f64)>> {
+        let mut connection = self.pool.acquire().await?;
+        let mut ts = since.duration_since(std::time::UNIX_EPOCH)?.as_secs() as i64;
+        ts = (ts / self.window) * self.window;
+        let records: Vec<(String, f64)> = sqlx::query_as(
+            "SELECT metric, SUM(value) FROM usage WHERE timestamp >= ? GROUP BY metric ;",
+        )
+        .bind(ts)
+        .fetch_all(&mut connection)
+        .await?;
+        Ok(records)
+    }
+
+    async fn slot(&mut self) -> Result<RRDSlotImpl> {
+        let connection = self.pool.acquire().await?;
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)?
+            .as_secs() as i64;
+        let ts = (now / self.window) * self.window;
+        RRDImpl::retain(self, ts).await?;
+        Ok(RRDSlotImpl {
+            connection,
+            key: ts,
+        })
+    }
+}
+
+impl RRDImpl {
+    /// new creates a new rrd database that uses sqlite as storage. if window or retention are 0
+    /// the function will return an RRDError. If retention is smaller then window the function will return an RRDError.
+    /// retention and window must be multiple of 1 minute.
+    pub async fn new<P: AsRef<Path>>(
+        path: P,
+        window: time::Duration,
+        retention: time::Duration,
+    ) -> Result<RRDImpl> {
+        if window.is_zero() {
+            return Err(Error::RRDError {
+                error: String::from("invalide window, can't be zero"),
```

Comment:
Shouldn't be this `Error::InvalidWindow` instead? creating a concrete error type then just return a generic error message does not make any sense.